### PR TITLE
IV Drip Buffs

### PR DESCRIPTION
**Pull request by VisVirific**
What began as a "simple" port ended up snowballing into this. I am severely sorry.

* Ports Eris' Medical Stand gas tank functionality and sprites. And immediately and intensely edits them to be more unique to Aurora.
* You can now knock IV stands over and trip on fallen IV stands, also bringing down whoever is attached via the mask.
* Radial Menu since IVs are very much interact-able now
* IV stands are now buildable and upgradable. Science can print their circuits, and also be bothered incessantly about upgrading them!
* Tweaks so that you can actually choose which body part you'll put the IV in, which is subjected to attaching delays and armor checks without any upgrades.

**New IV Sprites**

Left- Aurora | Middle - Old | Right- Eris

**Adding and removing things**

**Radial Menus**

**Knocking IVs down + trips**

Tripping only happens on **Harm Intent + Run**

**Body Part IV Choosing** (*14/03/2022 Edit: Outdated gif. Doesn't take in consideration the armor checks + delays)

### CODE REVIEW

**Comment by VisVirific** [comment]
!review

**Comment by alsoandanswer** [comment]
Is there still right-click functions? Important when the radials decide to lag out.

**Comment by VisVirific** [comment]
> Is there still right-click functions? Important when the radials decide to lag out.

Right clicks show the toggable functions. As for the radial menu lag, I don't think I've ever had a problem with it, hence I never realizing it was a possible thing

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:160]:
```suggestion
		if(gauge_pressure > TANK_IDEAL_PRESSURE )
```

Just use the actual define here, it isn't undefined in tanks.dm.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:163]:
```suggestion
			gauge_pressure = round((gauge_pressure/TANK_IDEAL_PRESSURE)*tank.gauge_cap)
```

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:170]:
```suggestion

```
Don't do this, the machinery subsystem will handle it.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:179]: I would prefer a time-since-last-squeaked to make sure this doesn't spam, rather than a prob.

It can have a prob as well if you want I guess.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:287]: Can you add tooltips to all the radial menu options too, like the robot chassis selection radial or the cooking radials.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:288]: It is better to define things in this case, e.g.

`#define IV_DRIP "IV drip"`

So the compiler will yell at you if you make a spelling mistake, since you're typing this string out in many places.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:367]: Pointless `else`, there are others like this in this PR to remove too.

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:170]: Surprisingly enough that was in the original IV code. I didn't know whether to remove it or not, so I decided to not touch it as I was re-organizing the code by the end. I'll remove it shortly

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:287]: Originally I intended to do so. Looks like I forgot to re-activate them in the final clean up. My bad

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:179]: Mm! I thought a prob() would be more convenient since it doesn't need to check for time it has last squeaked, but I'll definitely change it

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:160]: The problem is that it's saying it's undefined when I try to compile in VSCode

Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:160]: It's a define, it isn't a var of the tank object. 

All a define does is, when it compiles, the compiler finds every instance of `TANK_IDEAL_PRESSURE` in the code and pretends you wrote `1015`.

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:160]: That's the thing, my VSCode isn't compiling when I change it to the ``TANK_IDEAL_PRESSURE`` and keeps insisting it's an undefined var

I don't know if it's because the define set in the ``code/game/objects/items/weapons/tanks/tanks.dm`` itself over a proper ``code/__defines/tanks.dm`` file

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:160]: So, turns out the ``.../machinery/iv_drip.dm`` initializes earlier than the ``.../tanks/tanks.dm``. One of the fixes I could do is merely move the defines in the``.../tanks/tanks.dm`` to the ``.../__defines/misc.dm``, which allows my VSCode to compile it without having to create a new define file, or messing with the atmos.dm file

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:288]: I've been using ``[src]`` more than ``IV drip`` whenever I can for this code to avoid typos involving the name. There's really only 4 instances of using ``IV drip``, one of them being the name, the desc_info, and the MouseDrop. I really don't think it's needed in this case

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm]: > So, turns out the `.../machinery/iv_drip.dm` initializes earlier than the `.../tanks/tanks.dm`. One of the fixes I could do is merely move the defines in the`.../tanks/tanks.dm` to the `.../__defines/misc.dm`, which allows my VSCode to compile it without having to create a new define file, or messing with the atmos.dm file

That works.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:276]: There's no harm in using more defines than necessary, especially if you undef them after the codeblock. But this isn't critical, so I'll leave it up to you.

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:276]: Got it. I'll definitely keep it in mind for future code shenaniganry. Thank you!

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:297]: You should also check if they are wearing a voidsuit / armour / whatever, like syringes do. And maybe something in process for the same incase they put armour on while they have an IV needle in.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:185]: Some of the returns in the breather processing logic will prevent the IV drip from processing, when it should. Having the mask deactivate shouldn't skip the IV drip for that tick.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:249]: Not sure if it _needs_ to check the contents of the bag. It's an IV drip, not a reagent scanner. What if someone wants a coffee IV / something more nefarious? Or a medic wants a drug cocktail IV.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:261]: Could use a wait timer here too for the sound effect?

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:266]: So as I understand it this message only happens once, and is only reset when you toggle? I think this one should repeat every 10s or so because it has the potential to kill people if ignored as it doesn't stop the blood being taken.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:274]:
```suggestion
	if(in_range(src, usr) && ishuman(over_object) && in_range(over_object, src))
```

Might aswell use in_range() for both checks, since all it does is `get_dist(thing, other_thing) <= 1`

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:402]: Still got a pointless else here.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:438]: Why is the phoron tank, specifically, harder to add to tipped IVs + not securable? Is it supposed to be heavier or more awkward or something?

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:462]: I think this check should instead be done with a `/proc/attack_ai()` or `/proc/ai_can_interact()` override / something.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:561]: If we're not living, then we already failed the check on L558 because human and silicon are both `/mob/living` subtypes.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:665]:
```suggestion
	if (!(user in viewers(2)) && user!=src.loc)
```

Better to use `viewers` than `view` here because byond optimizes it a bit and only returns mobs from `viewers`.

Also what does the second part of this check do? What situation would cause the IV drip to be inside the mob examining it?

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:249]: The toggle stop is only for blood/saline and only if the person's blood volume is at 100%. Otherwise the IV will inject anything and stuff

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:261]: Ah, that was from the original IV code which I didn't mess too much with. Will do so!

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:402]: FUCK

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:438]: It's mostly because of spriting issues. It clipped out of frame whenever I put the phoron on the correct position and looked really weird in turn

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:665]: I see, I'll change it shortly.

Also, I have no idea at all what the second part is about since that whole string ``if(!(user in view(2)) && user != src.loc`` is part of the original IV code which I didn't mess with. I'll clear it up though!

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:266]: Yeah, pretty much. I wasn't entirely sure of the original intent so I didn't mess with it, but I'll add a visual overlay as well as the repeating buzz

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:185]: Yeah, I realized a bit late regarding how some of the previous checks would do that for the injecting/intaking mode for the IV when someone was wearing the mask and changed it accordingly, mostly by moving the checks and returns to only happen if the valve is open. The returns are still there to make sure it won't continue to proceed and be fully certain it'll prevent any strange behavior. The current iteration of the checks/returns is the best I managed to get to

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:307]: The latter sounds like a nightmare that it's one of the few things I'm looking forward to be ordered to do, while the former I'm on the fence on since armor/voidsuit checks can be The Thing that either kills the patient or not, especially if they're wearing a RIG/hardsuit. Unless I can come up with a good solution to this, I'm not coding in that check

Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:307]: Try not to mark things as resolved if they're not actually resolved, so I / other reviewers / maintainers can easily see them.

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:189]: Same here. I can help correct this if you want, just have to ask (same for any of my review comments).

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:307]: I see. My bad about the resolved thing.

I'll see what I can do about this thing in particular. Ideally I really do not want to make cases with RIGs even more impossible to treat and just stare at them die. I'll probably just add an injection timer thing and make the IVs upgradable via Science so they can retain their bypass

File: code/game/machinery/iv_drip.dm
Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:189]: I really don't know if that's even possible without gutting out the entire code for the Breather/Attached again, or adding yet another check to just jump lines (if that's even a thing), or some ``proc/attached_in()`` nonsense that'd make this even more confusing

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:307]: You could ask a maintainer about what their preferred solution might be.
A little timer while you "find an injection port" could work?

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:307]: Oh yeah, the armor check is already coded and working. I've been staring at the returns for the breather thing for the majority of today

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:189]: You can do something like this so the returns only return from the breather OR attached portions rather than the entire machinery proc.

[Example Code For You](https://github.com/mikomyazaki/Aurora.3/blob/3784724babe29f3f5949439fcdab25fca7744a81/code/game/machinery/iv_drip.dm#L169)

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:189]: So the answer really was the fucking nonsense, oh my god

Comment (mikomyazaki) [reply code/game/machinery/iv_drip.dm:189]: Is it nonsense? You could also make it work by rearranging the ifs + using more elses. But I think that gets more tangled.

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:189]: Nonsense because at the time I considered it, I was already stressed and began to over-complicate the possible solution into something I thought as unreasonable. Though now I'll definitely apply that

Comment (VisVirific) [reply code/game/machinery/iv_drip.dm:212]: And done! Thank you very much!

File: code/game/machinery/iv_drip.dm
Comment (mikomyazaki) [inline code/game/machinery/iv_drip.dm:186]: This is a little nitpicky, but I think it is better to have just `if(breather)` here. Breather can only be `null` or `/mob/living/carbon/human` -- And if, somehow, it ever isn't then we will want this proc to runtime and complain. If we check the `istype` then we suppress those useful errors. Same on attached_process.

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

**Review by mikomyazaki — APPROVED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Pull request merged**
